### PR TITLE
Add safeIndex

### PR DESCRIPTION
**Pull request by novemberkilo**
! @nhibberd @erikd-ambiata

Pull across `safeIndex` from `Deckard.Util`

### CODE REVIEW

**Comment by erikd-ambiata** [comment]
:100: on :green_heart: !!

File: src/P/List.hs [hunk 1]
```diff
@@ -45,3 +46,11 @@ lastMaybe = listToMaybe . reverse
 -- | count the number of elements satisfying a predicate in a list
 count :: (a -> Bool) -> [a] -> Int
 count predicate = length . filter predicate
+
+safeIndex :: [a] -> Int -> Maybe a
+safeIndex xs i
```
Comment (charleso) [inline src/P/List.hs]: Wondering if the args should be flipped. (eg. `Int -> [a] -> Maybe a`). That's better for partial application, but not so much infix notation. (I don't have an opinion on it though)

Comment (novemberkilo) [inline src/P/List.hs]: fwiw it follows the same convention as `(!!)` http://hackage.haskell.org/package/base-4.9.0.0/docs/Prelude.html#v:-33--33-

File: src/P/List.hs [hunk 2]
```diff
@@ -45,3 +46,11 @@ lastMaybe = listToMaybe . reverse
 -- | count the number of elements satisfying a predicate in a list
 count :: (a -> Bool) -> [a] -> Int
 count predicate = length . filter predicate
+
+safeIndex :: [a] -> Int -> Maybe a
```
Comment (nhibberd) [inline src/P/List.hs]: What are peoples thoughts on `(!?)` with the args flipped?

File: src/P/List.hs (hunk 1 above)
Comment (erikd-ambiata) [inline src/P/List.hs]: @charleso I wrote it to have the same arg order as `!!` and used it as an infix operator.

File: src/P/List.hs (hunk 2 above)
Comment (erikd-ambiata) [inline src/P/List.hs]: I'd be fine with that.

Comment (jystic) [inline src/P/List.hs]: I'd be -1 for `(!?)` in P, perfectly fine if it was X though

Comment (erikd-ambiata) [inline src/P/List.hs]: Actually, why not provide both `safeIndex` and `(!?)` and implement the second as:
```
(!?) :: [a] -> Int -> Maybe a
(!?) = safeIndex
```
People are more likely to use the `(!?)` operator, but when they go looking for something they are more likely to hoogle for `safeIndex`. I'm pretty sure thats how @novemberkilo found that function in Deckard.

I'm pretty

Comment (erikd-ambiata) [inline src/P/List.hs]: @jystic why not in `P`?

**Comment by thumphries** [comment]
Not going to contribute to bikeshedding on the operator, but I'd use this.

Almost everything in `safe` should be in P in my opinion

https://hackage.haskell.org/package/safe

**Comment by jystic** [comment]
I could be convinced if it was more general, like over Foldable perhaps, otherwise I think maybe we should make `X.Data.List`

fwiw lens (of course) has a crazy type class version of [this](http://hackage.haskell.org/package/lens-4.14/docs/Control-Lens-At.html#t:Ixed) idea that works efficiently on just about anything

**Comment by novemberkilo** [comment]
IRL discussion with @jystic and @charleso has convinced me that this is not appropriate for `P` 

Closing ...

**Pull request closed without merging**